### PR TITLE
Add batched featurization logic to reduce memory usage in repair model

### PR DESCRIPTION
**Pull request by jonmio**
WIP - PR is missing documentation.

Currently, featurization is done on all cells, and the resulting feature tensors are stored in memory. The resulting feature tensors are very large relative to size of the original dataset, potentially resulting in memory issues.

This PR introduces changes to address these memory concerns. Torch DataLoader is used to iterate over batches of the training/inference data, and featurization happens only during training/inference of the example.

### CODE REVIEW

File: repair/featurize/freqfeat.py [hunk 1]
```diff
@@ -12,24 +12,25 @@ def specific_setup(self):
         total, single_stats, pair_stats = self.ds.get_statistics()
         self.total = total
         self.single_stats = single_stats
+        # The query results used to featurize the dataset
```
Comment (richardwu) [inline repair/featurize/freqfeat.py]: Can you add a comment here denoting the structure of `featurization_query_results` e.g.
```
List of tuples (vid, attribute, ||| separated domain values)
```

File: repair/featurize/constraintfeat.py [hunk 2]
```diff
@@ -46,30 +45,50 @@
...
         self.constraints = self.ds.constraints
         self.init_table_name = self.ds.raw_data.name
+        self.featurization_query_results = self._get_featurization_query_results()
+        # List[Dict[str, Dict[str, int]]] where the key of the dict is the vid
+        # and the value is a dict whtere the key is the value_id and the value
+        # is the number of violations.
+        self.featurization_maps = self.build_featurization_maps(self.featurization_query_results)
+
+    # TODO(jmio): potential memory issues
+    def build_featurization_maps(self, queries):
```
Comment (richardwu) [inline repair/featurize/constraintfeat.py:64]: Can you add a docstring here specifying the structure of `queries`? Also above `self.featurization_query_results`.

File: repair/featurize/constraintfeat.py [hunk 3]
```diff
@@ -46,30 +45,50 @@
...
+        featurization_maps = []
+        for query in queries:
+            featurization_map = {}
+            for result in query:
+                vid = int(result[0])
+                val_id = int(result[1]) - 1
+                feat_val = float(result[2])
+                if vid not in featurization_map:
+                    featurization_map[vid] = {}
+                if val_id not in featurization_map[vid]:
```
Comment (richardwu) [inline repair/featurize/constraintfeat.py]: This block shouldn't be necessary since you are always writing a value `feat_val` into `featurization_map[vid][val_id]`.

File: repair/featurize/constraintfeat.py [hunk 4]
```diff
@@ -46,30 +45,50 @@
...
-    return tensor
-
-
 class ConstraintFeaturizer(Featurizer):
     def specific_setup(self):
         self.name = 'ConstraintFeaturizer'
         self.constraints = self.ds.constraints
         self.init_table_name = self.ds.raw_data.name
+        self.featurization_query_results = self._get_featurization_query_results()
+        # List[Dict[str, Dict[str, int]]] where the key of the dict is the vid
```
Comment (richardwu) [inline repair/featurize/constraintfeat.py:58]: Can you add a comment about how each `dict` in the list corresponds to one denial constraint query?

File: repair/featurize/constraintfeat.py [hunk 5]
```diff
@@ -46,30 +45,50 @@
...
+        tensors = []
+        for featurization_map in self.featurization_maps:
+            tensor = torch.zeros(self.classes, 1)
+            if vid in featurization_map:
+                for val_id in featurization_map[vid]:
+                    violation_count = featurization_map[vid][val_id]
+                    tensor[val_id][0] = violation_count
+            tensors.append(tensor)
+        combined = torch.cat(tensors, dim=1)
+        combined = F.normalize(combined, p=2, dim=0)
```
Comment (richardwu) [inline repair/featurize/constraintfeat.py]: Hmm I'm not too sure why normalization was added here since we do (optionally) normalization again afterwards. Can you remove this for now?

File: repair/featurize/constraintfeat.py [hunk 6]
```diff
@@ -46,30 +45,50 @@
...
+                    featurization_map[vid] = {}
+                if val_id not in featurization_map[vid]:
+                    featurization_map[vid][val_id] = {}
+                featurization_map[vid][val_id] = feat_val
+            featurization_maps.append(featurization_map)
+        return featurization_maps
+
+    def gen_feat_tensor(self, vid):
+        tensors = []
+        for featurization_map in self.featurization_maps:
```
Comment (richardwu) [inline repair/featurize/constraintfeat.py:93]: Maybe add a comment here along the lines of "iterate through memoized feature map for every DC constraint..."

File: repair/featurize/initattrfeat.py [hunk 7]
```diff
@@ -6,31 +6,28 @@
...
-
-
 class InitAttrFeaturizer(Featurizer):
     def specific_setup(self):
         self.name = 'InitAttrFeaturizer'
         self.all_attrs = self.ds.get_attributes()
         self.attr_to_idx = self.ds.attr_to_idx
         self.total_attrs = len(self.ds.attr_to_idx)
+        # The query results used to featurize the dataset
+        self.featurization_query_results = self._get_featurization_query_results()
```
Comment (richardwu) [inline repair/featurize/initattrfeat.py:33]: Same thing here with a comment on the structure of query results.

File: repair/learn/learn.py [hunk 8]
```diff
@@ -88,9 +92,7 @@ def __init__(self, env, feat_info, output_dim, bias=False):
         self.model = TiedLinear(self.env, feat_info, output_dim, bias)
         self.featurizer_weights = {}
 
-    def fit_model(self, X_train, Y_train, mask_train):
-        n_examples, n_classes, n_features = X_train.shape
-
+    def fit_model(self, training_data):
```
Comment (richardwu) [inline repair/learn/learn.py:94]: Can you add a docstring here specifying that `training_data` is a `torch.utils.data.Dataset`?

File: repair/featurize/featurized_dataset.py [hunk 9]
```diff
@@ -111,32 +93,51 @@ def generate_var_mask(self):
             var_to_domsize[vid] = max_class
         return mask, var_to_domsize
 
-    def get_tensor(self):
-        return self.tensor
-
+    # Returns an instance of TorchFeaturizedDataset for the training data
     def get_training_data(self):
```
Comment (richardwu) [inline repair/featurize/featurized_dataset.py]: Maybe rename this to`get_training_dataset` and below as `get_infer_dataset`.

**Review by richardwu — CHANGES_REQUESTED** [review]
Good job Jon, I think you're on the right track here to bound memory usage by some upper bound.

One idea to reduce runtime on epochs after the first is to cache the tensors up to some fixed # of VIDs (such that size of tensor is ~ X GB where X is a parameter), then dump it to disk (`pickle.dumps` is probably the best option here into some `.tmp` file vs dumping into Postgres). You could keep an internal map from VID to filename to keep track of which tensor is in which dump file. You would essentially do an eager load the next time the first VID from a dump file is looked up.

One issue here is that we do end up `shuffle`ing the batches during training which would ultimately cause random access between dump files. You can try disabling `shuffle` for now and test if this improves runtime by much.

An easy hack to see if this will help is to just cache the tensor after the first epoch and do a lookup into the tensor.

File: repair/featurize/constraintfeat.py (hunk 5 above)
Comment (jonmio) [reply repair/featurize/constraintfeat.py]: Done

**Comment by jonmio** [comment]
Added documentation which address most of Richard's comments. 

Also, ran some experiments with memory and runtime profiling. On the full hospital dataset, we can reduce memory usage in training from 900MB to 230MB by using an iterator with featurization_batch_size of 10. However, each epoch in training takes about ~20s longer because we re-featurize on each epoch.

Precision and recall haven't been affected.

Regarding caching after the first epoch, I've also discussed this with Mina and we plan on experimenting with it. One workaround he suggested to the shuffling issue is to shuffle within batches.

File: repair/featurize/constraintfeat.py [hunk 10]
```diff
@@ -46,31 +45,71 @@
...
+        tensors = []
+        # Iterate over each DC and check whether it was violated by the cell.
+        for featurization_map in self.featurization_maps:
+            tensor = torch.zeros(self.classes, 1)
+            if vid in featurization_map:
+                for val_id in featurization_map[vid]:
+                    violation_count = featurization_map[vid][val_id]
+                    tensor[val_id][0] = violation_count
+            tensors.append(tensor)
+        combined = F.normalize(torch.cat(tensors, dim=1), p=2, dim=0)
```
Comment (jonmio) [inline repair/featurize/constraintfeat.py:100]: Richard mentioned this shouldn't be necessary, but including this for now, in case it's affecting recall

**Review by minafarid — APPROVED** [review]
